internal/js/modules/k6/browser/common/page.go

```diff
@@ -210,6 +211,99 @@ func NewRouteHandler(
 
 type RouteHandlerCallback func(*Route) error
 
+type ResponseEventHandler struct {
+	mu            sync.RWMutex
+	activeWaiters map[int64]*responseWaiter
+	nextWaiterID  int64
+}
+
+type responseWaiter struct {
+	matcher      URLMatcher
+	responseChan chan *Response
+	ctx          context.Context
+	cancel       context.CancelFunc
+}
+
+func NewResponseEventHandler() *ResponseEventHandler {
+	return &ResponseEventHandler{
+		activeWaiters: make(map[int64]*responseWaiter),
+	}
+}
+
+func (reh *ResponseEventHandler) addWaiter(waiter *responseWaiter) int64 {
+	id := reh.generateWaiterID()
+	reh.mu.Lock()
+	defer reh.mu.Unlock()
+
+	reh.activeWaiters[id] = waiter
+	return id
+}
+
+func (reh *ResponseEventHandler) removeWaiter(id int64) {
+	reh.mu.Lock()
+	defer reh.mu.Unlock()
+	delete(reh.activeWaiters, id)
+}
+
+func (reh *ResponseEventHandler) processResponse(response *Response) {
+	if response == nil {
+		return
+	}
+
+	reh.mu.RLock()
+	defer reh.mu.RUnlock()
+	waitersToNotify := make([]*responseWaiter, 0)
+
+	for _, waiter := range reh.activeWaiters {
+		// Check if the context is still active before processing
+		select {
+		case <-waiter.ctx.Done():
+			continue
+		default:
+		}
+
+		matched, err := waiter.matcher(response.URL())
+		if err == nil && matched {
+			waitersToNotify = append(waitersToNotify, waiter)
+		}
+	}
```

Comment:
Why do we need to put these waiters in a slice? Can you loop once without the slice?